X11R6/xc/doc/man/Xi/XSetDVal.man
.\"
.\" Copyright ([\d,\s]*) by Hewlett-Packard Company, Ardent Computer, 
.\" 
.\" Permission to use, copy, modify, distribute, and sell this documentation 
.\" for any purpose and without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\" Ardent, and Hewlett-Packard make no representations about the 
.\" suitability for any purpose of the information in this document.  It is 
.\" provided \`\`as is'' without express or implied warranty.
.\" 
.\" $XConsortium: XSetDVal.man,v 1.6 94/06/04 17:32:38 rws Exp $
.ds xL Programming with Xlib
.TH XSetDeviceValuators 3X11 "Release 6" "X Version 11" "X FUNCTIONS"
.SH NAME
XSetDeviceValuators \- initialize the valuators on an extension input device
.SH SYNTAX
XSetDeviceValuators\^(\^\fIdisplay\fP, \fIdevice\fP\^, \fIvaluators\fP\^, \fIfirst_valuator\fP\^, \fInum_valuators\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      XDevice *\fIdevice\fP\^; 
.br
      int *\fIvaluators\fP\^; 
.br
      int \fIfirst_valuator\fP\^; 
.br
      int \fInum_valuators\fP\^; 
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I device
Specifies the device whose valuators are to be initialized.
.TP 12
.I valuators
Specifies a pointer to an array of integer values to be used to initialize the
device valuators.
.TP 12
.I first_valuator
Specifies the first valuator to be set.  Valuators are numbered beginning with
zero.
.TP 12
.I num_valuators
Specifies the number of valuators to be set.
.SH DESCRIPTION
The
\fIXSetDeviceValuators\fP
request sets the current values of the valuators of an input device.
The valuators in the range
first_valuator to (first_valuator + num_valuators)
are set to the specified values.  Valuators are numbered beginning with 0.
Not all input devices support initialization of valuator values.  If this
request is made to a device that does not support valuators initialization,
a \fIBadMatch\fP error will occur.
.LP
If the request succeeds, a status of Success is returned.  If another
client has the device grabbed, a status of AlreadyGrabbed is returned.
.LP
\fIXSetDeviceValuators\fP
can generate a \fIBadLength\fP, \fIBadDevice\fP , \fIBadMatch\fP, or 
\fIBadValue\fP error.
.SH DIAGNOSTICS
.TP 12
\fIBadDevice\fP
An invalid device was specified.  The specified device does not exist or has 
not been opened by this client via \fIXOpenInputDevice\fP.
.TP 12
\fIBadMatch\fP
This error may occur if an \fIXSetDeviceValuators\fP request is made 
specifying a device that has no valuators and reports no axes of motion,
or if such a request is made specifying a device that does not support
valuator initialization.
.TP 12
\fIBadValue\fP
An invalid first_valuator or num_valuators values was specified.
.SH "SEE ALSO"
.br
\fI\*(xL\fP


X11R6/xc/lib/Xi/XSetDVal.c
/* $XConsortium: XSetDVal.c,v 1.3 94/04/17 20:18:12 rws Exp $ */

/************************************************************

Copyright (c) 1989  X Consortium

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of the X Consortium shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from the X Consortium.

Copyright (c) 1989 by Hewlett-Packard Company, Palo Alto, California.

			All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the name of Hewlett-Packard not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

HEWLETT-PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HEWLETT-PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

********************************************************/

/***********************************************************************
 *
 * XSetDeviceValuators - Set the value of valuators on an extension input
 * device.
 *
 */

#include "XI.h"
#include "XIproto.h"
#include "Xlibint.h"
#include "XInput.h"
#include "extutil.h"

int
XSetDeviceValuators (dpy, dev, valuators, first_valuator, num_valuators)
    register Display 	*dpy;
    XDevice 		*dev;
    int			*valuators;
    int			first_valuator;
    int			num_valuators;
    {       
    xSetDeviceValuatorsReq 		*req;
    xSetDeviceValuatorsReply 	rep;
    XExtDisplayInfo *info = (XExtDisplayInfo *) XInput_find_display (dpy);

    LockDisplay (dpy);
    if (_XiCheckExtInit(dpy, XInput_Add_XSetDeviceValuators) == -1)
	return (NoSuchExtension);

    GetReq(SetDeviceValuators,req);		
    req->reqType = info->codes->major_opcode;
    req->ReqType = X_SetDeviceValuators;
    req->deviceid = dev->device_id;
    req->first_valuator = first_valuator;
    req->num_valuators = num_valuators;
    req->length += num_valuators;

    /* note: Data is a macro that uses its arguments multiple
       times, so "nvalues" is changed in a separate assignment
       statement */

    num_valuators <<= 2;
    Data (dpy, (char *) valuators, num_valuators);

    (void) _XReply (dpy, (xReply *) &rep, 0, xTrue);
    UnlockDisplay(dpy);
    SyncHandle();
    return (rep.status);
    }

